### PR TITLE
Refine yawollet health check in keepailved

### PR DESCRIPTION
**Pull request by maboehm**
Currently, the keepalived priority is influenced by a `track_file` which is supposed to represent that the LBM matches the latest revision. This allows for a rolling update, since all older instances will have a lower priority.

However, the track file is not always removed correctly, for example if the yawollet cannot reach the API-Server at all. So this change replaces the file-based check with a health endpoint, which is checked by keepalived via `track_script`. This now covers 3 conditions:
- is the yawollet running correctly (if now wget gets no response/timeouts)
- are the informers/caches synced
- is the LBM the current revision
- Is the LBM condition heartbeat up-to-date.

The last check is essentially a check if we can reach the API-Server. We could call the `/healthz` endpoint, but that might cause unnecessary load. By checking the conditions that the yawollet should be writing itself, we check both that we can write and read (watch) successfully from the API-Server without causing additional calls.

### CODE REVIEW

File: internal/healthz/heartbeat.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+)
+
+// NewHeartbeatHeathz returns a new healthz.Checker that checks the condition
+// heartbeat of the LBM that are written by the yawollet. This way we can check
+// if we have a healthy connection to the API-Server, both writing (updating the
+// heartbeats), and reading (retrieving the updated heartbeats), without
+// additional requests to the API Server.
+func NewHeartbeatHeathz(ctx context.Context,
```
Comment (timebertt) [inline internal/healthz/heartbeat.go:21]:
```suggestion
func NewHeartbeatHealthz(ctx context.Context,
```

File: internal/healthz/heartbeat.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	)
+	return func(_ *http.Request) error {
+		if err := reader.Get(ctx, lbmKey, lbm); err != nil {
+			return fmt.Errorf("failed getting LoadBalancerMachine %q", lbmKey)
+		}
+		expiration := metav1.Time{Time: time.Now().Add(-expiration)}
+
+		ok, reason := helper.AreRelevantConditionsMet(lbm, expiration, false)
+		if !ok {
+			return fmt.Errorf("not all relevant conditions are met, reason: %q", reason)
```
Comment (timebertt) [inline internal/healthz/heartbeat.go:40]:
```suggestion
			return fmt.Errorf("not all relevant conditions are met, reason: %s", reason)
```

File: cmd/yawollet/main.go [hunk 3]
```diff
@@ -226,17 +229,37 @@ func main() {
...
 		os.Exit(1)
 	}
+
+	// The readyz checks are used by keepalived for increasing the priority of machines that run a healthy yawollet for
+	// the latest LoadBalancer revision (i.e., when the machine belongs to the current LoadBalancerSet).
 	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
 		setupLog.Error(err, "unable to set up ready check")
 		os.Exit(1)
 	}
+	if err := mgr.AddReadyzCheck("loadbalancer-heartbeat", yawolhealthz.NewHeartbeatHeathz(
```
Comment (timebertt) [inline cmd/yawollet/main.go:245]: Let's move this one down to the other loadbalancer check

File: cmd/yawollet/main.go [hunk 4]
```diff
@@ -226,17 +229,37 @@ func main() {
...
+		managerCtx, mgr.GetCache(), 2*requeueDuration, namespace, loadbalancerMachineName)); err != nil {
+		setupLog.Error(err, "unable to set up API server ready check")
```
Comment (timebertt) [inline cmd/yawollet/main.go:247]:
```suggestion
		setupLog.Error(err, "unable to set up LoadBalancer heartbeat ready check")
```

**Review by timebertt — COMMENTED** [review]
Just a few nits

**Review by timebertt — DISMISSED** [review]
Nice!

**Review by MichaelEischer — APPROVED** [review]

**Review by timebertt — DISMISSED** [review]

File: internal/helper/loadbalancermachine.go [hunk 5]
```diff
@@ -305,6 +308,11 @@ write_files:
   owner: root:root
   path: /etc/keepalived/keepalived.conf
   permissions: '0644'
+- encoding: b64
+  content: ` + yawolletHealthCheckScriptBase64 + `
+  owner: root:root
+  path: ` + YawolletHealthCheckScript + `
+  permissions: '0655'
```
[the remark this answers is not in the document]
Comment (timebertt) [reply internal/helper/loadbalancermachine.go:316]: Good catch, this was an oversight. Fixed

**Review by timebertt — APPROVED** [review]

**Pull request merged**